tools/ci_build/build.py

```diff
@@ -888,6 +898,48 @@ def generate_build_tree(cmake_path, source_dir, build_dir, cuda_home, cudnn_home
                 "-DCMAKE_CXX_COMPILER=" + compilers[1]
             ]
 
+    if args.build_wasm:
+        # install emscripten if not exist.
+        # since emscripten doesn't support file packaging required for unit tests,
+        # need to apply patch with the specific version of emscripten.
+        # once patch is committed to emsdk repository, this must be replaced with 'latest'.
+        emsdk_version = "2.0.13"
+
+        emsdk_dir = os.path.join(cmake_dir, "external", "emsdk")
+        emsdk_file = os.path.join(emsdk_dir, "emsdk.bat") if is_windows() else os.path.join(emsdk_dir, "emsdk")
+        emsdk_version_file = os.path.join(emsdk_dir, "upstream", "emscripten", "emscripten-version.txt")
+        emscripten_cmake_toolchain_file = os.path.join(emsdk_dir, "upstream", "emscripten", "cmake", "Modules",
+                                                       "Platform", "Emscripten.cmake")
+
+        if os.path.exists(emsdk_version_file):
+            with open(emsdk_version_file) as f:
+                emsdk_version_data = json.load(f)
+            emsdk_version_match = isinstance(emsdk_version_data, str) and emsdk_version_data == emsdk_version
+        if not os.path.exists(emscripten_cmake_toolchain_file) or not emsdk_version_match:
+            print("Installing emsdk...")
+            run_subprocess([emsdk_file, "install", emsdk_version], cwd=emsdk_dir)
+        print("Activating emsdk...")
+        run_subprocess([emsdk_file, "activate", emsdk_version], cwd=emsdk_dir)
+
```

Comment:
Would have expected installation be done elsewhere prior to calling generate_build_tree (which is ridiculously long enough already).

Same goes for setting up testing etc - doesn't really belong in generate_build_tree.